src/deskolem.ml

```diff
@@ -3,186 +3,216 @@ open! Lplib
-            assert (not (Bindlib.binder_occur b));
-            Term.add_args (mk_Symb imp) [unProofA; unProof sign (Bindlib.unbind b |> snd)]
-        with Not_ProofS _ ->
-            match a with
-            |Appl(Symb({sym_name = "τ"; _}), a') ->
-                let forall = Extra.StrMap.find "forall" !(sign.sign_builtins) in
-                Term.add_args (mk_Symb forall) [a'; unProof sign (Bindlib.unbind b |> snd)]
-            |_                                  ->
-                raise (Not_Proof t))
+    |Appl(Symb({sym_name = "ϵ"; _}), t')    -> t'
+    |Prod(a, b)                             ->
+        let cfg = get_config sign in
+        (
+            try
+                (* get the proposition in [a]. *)
+                let prop_a = try get_prop sign a with Not_Proof _ -> 
+                    raise (Not_ProofS a) in
+                (* check if the variable that has the type [a] is not in [b]. *)
+                assert (not (Bindlib.binder_occur b));
+                (* return the proposition of [a] => the proposition of [b]. *)
+                Term.add_args (mk_Symb cfg.symb_imp) 
+                    [prop_a; get_prop sign (Bindlib.unbind b |> snd)]
+            with Not_ProofS _ ->
+                match a with
+                (* check if the left part of the implication is a variable 
+                    declaration. *)
+                |Appl(Symb({sym_name = "τ"; _}), a') ->
+                    (* return that variable => the proposition of [b]. *)
+                    Term.add_args (mk_Symb cfg.symb_forall) 
+                        [a'; get_prop sign (Bindlib.unbind b |> snd)]
+                |_                                  ->
+                    raise (Not_Proof t)
+        )
 
-    |Abst(_, _) -> Common.Console.out 4 "ABST NOT PROOF : %a@." Print.pp_term t; raise (Not_Proof t)
-    |_                                          ->
-        Common.Console.out 4 "Other cases NOT PROOF : %a@." Print.pp_term t; raise (Not_Proof t)
+    |Abst(_, _)                             -> raise (Not_Proof t)
+    |_                                      -> raise (Not_Proof t)
 
-(** [unProofCheck t] returns true if the proposition is a proof. *)
-let rec unProofCheck : Sign.t -> term -> bool = fun sign t ->
-    Common.Console.out 4 "unProof [%a]@." Print.pp_term t;
+(** [check_epsilon t] returns true if the proposition is a proof. *)
+let rec check_epsilon : term -> bool = fun t ->
     match unfold t with
     |Appl(Symb({sym_name = "ϵ"; _}), _) -> true
-    |Prod(a, b) ->
+    |Prod(a, b)                         ->
         (try
-            if not (unProofCheck sign a) then
+            if not (check_epsilon a) then
                 raise (Not_ProofS a)
             else
-                unProofCheck sign (Bindlib.unbind b |> snd)
+                check_epsilon (Bindlib.unbind b |> snd)
         with Not_ProofS _ ->
             match a with
             |Appl(Symb({sym_name = "τ"; _}), _) ->
```

Comment:
This is not correct to test symbol name equality. You should use Term.is_symb cfg.symb_... instead.